### PR TITLE
feat: Watsonx

### PR DESCRIPTION
**Pull request by micaelakaplan**
WatsonX integration example

### CODE REVIEW

File: label_studio_ml/examples/watsonx_llm/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+
+
+# Integrate WatsonX to Label Studio
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/README.md:3]: can we add expected project XML labeling configuration to this section?

File: label_studio_ml/examples/watsonx_llm/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+WatsonX offers a suite of machine learning tools, including access to many LLMs, prompt
+refinement interfaces, and datastores via WatsonX.data. When you integrate WatsonX with Label Studio, you get 
+access to these models and can automatically keep your annotated data up to date in your WatsonX.data tables. 
+
+To run the integration, you'll need to pull this repo and host it locally or in the cloud. Then, you can link the model 
+to your Label Studio project under the `models` section in the settings. To use the WatsonX.data integration, 
+set up a webhook in settings under `webhooks` by using the following structure for the link:
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/README.md:10]: Please give the link to the Webhook's documentation here

File: label_studio_ml/examples/watsonx_llm/_wsgi.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+from werkzeug.middleware.dispatcher import DispatcherMiddleware
+from flask import Flask
+
+from data_wsgi import application as data
+import model_wsgi as model
+
+app = Flask(__name__)
+
+app.wsgi_app = DispatcherMiddleware(model.application.wsgi_app, {
+    '/data': data.wsgi_app
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/_wsgi.py:10]: please add comment what does it extend

File: label_studio_ml/examples/watsonx_llm/data_transfer_app.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+logger = logging.getLogger(__name__)
+_server = Flask(__name__)
+
+
+def init_app():
+    return _server
+
+
+# @_server.route('/')
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/data_transfer_app.py:18]: please remove commented code

File: label_studio_ml/examples/watsonx_llm/data_transfer_app.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    input_request = request.json
+    action = input_request["action"]
+    annotation = input_request["annotation"]
+    task = input_request["task"]
+
+    # Connect to Label Studio
+    client = connect_ls()
+    data = get_data(annotation, task, client)
+
+    # Then, connect to WatsonX.data via prestodb
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/data_transfer_app.py:46]: does it work with default "None" values? if not, defaults should be provided explicitly

File: label_studio_ml/examples/watsonx_llm/data_transfer_app.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+def connect_ls():
+    try:
+        client = LabelStudio(
+            base_url = os.getenv("LABEL_STUDIO_URL"),
+            api_key=os.getenv("LABEL_STUDIO_API_KEY")
+        )
+    except Exception as e:
+        print(e)
+    return client
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/data_transfer_app.py:100]: nit functions should be separated by newlines

File: label_studio_ml/examples/watsonx_llm/data_transfer_app.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                value = int(value)
+
+            if isinstance(value, str):
+                value = value.strip("\"")
+            info.update({key: value})
+
+        print(f"INFO {info}")
+        return info
+    except Exception as e:
+        print(e)
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/data_transfer_app.py:141]: must be more descriptive message here, including the traceback `traceback.format_exc()`

File: label_studio_ml/examples/watsonx_llm/data_transfer_app.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+    table_info_keys = sorted(table_info.keys())
+    table_info_keys.insert(0, table_info_keys.pop(table_info_keys.index("ID")))
+    nl = ",\n"
+    strings = [f"{key} {table_info[key]}" for key in table_info_keys]
+    table_create = f"""
+    CREATE TABLE IF NOT EXISTS {table} ({nl.join(strings)})
+
+    """
+    print(table_create)
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/data_transfer_app.py:162]: all print() should be replaced with
```
import logging
logger = ogging.getLogger(__name__)
...
logger.debug(...)  # or logger.info()

File: label_studio_ml/examples/watsonx_llm/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
+version: "3.8"
+
+services:
+  watsonx_llm:
+    container_name: watsonx_llm
+    image: humansignal/ml-backend:v0
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/docker-compose.yml:6]: watsonx_llm should be in the image name

File: label_studio_ml/examples/watsonx_llm/model.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        """
+        if self.MODEL_TYPE not in ModelTypes.__members__:
+            raise Exception(f"WATSONX_MODELTYPE must be one of {[model for model in ModelTypes.__members__]}")
+
+        # If you want to specify model parameters, you can do so using this empty parameters dictionary
+        parameters = {
+            GenParams.MAX_NEW_TOKENS: 50,
+        }
+
+        self.model = ModelInference(
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/model.py:52]: does this code assume reinitialization? how costly is this run? because `setup()` is called quite often, it may slow down predictions if reinitializing every time

File: label_studio_ml/examples/watsonx_llm/model.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+            model_id=ModelTypes[self.MODEL_TYPE],
+            params=parameters,
+            credentials=self.WATSONX_CREDENTIALS,
+            project_id=self.PROJECT_ID
+        )
+        self.set("model_version", f'{self.__class__.__name__}-v0.0.1')
+
+    def _ocr(self, image_url):
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/model.py:60]: to avoid copypaste, can we move this code into tools? we can do this in a separate PR though

File: label_studio_ml/examples/watsonx_llm/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,7 @@
+ibm-watsonx-ai
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/requirements.txt:1]: let's pin versions

File: label_studio_ml/examples/watsonx_llm/model.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+    MODEL_TYPE = os.getenv('WATSONX_MODELTYPE')
+    PROMPT_PREFIX = os.getenv('PROMPT_PREFIX', "prompt")
+    USE_INTERNAL_PROMPT_TEMPLATE = bool(int(os.getenv("USE_INTERNAL_PROMPT_TEMPLATE", 1)))
+    DEFAULT_PROMPT = os.getenv('DEFAULT_PROMPT')
+    PROMPT_TEMPLATE = os.getenv('PROMPT_TEMPLATE', '**Source Text**:\n\n"{text}"\n\n**Task Directive**:\n\n"{prompt}"')
+    PROMPT_TAG = "TextArea"
+    SUPPORTED_INPUTS = ("Image", "Text", "HyperText", "Paragraphs")
+
+
+
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/model.py:40]: nit: 2 lines between variables and methods block

File: label_studio_ml/examples/watsonx_llm/test_api.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+              smart="false"
+              placeholder="Generated response will appear here..."
+    />
+    <Header value="Overall response quality:"/>
+    <Rating name="rating" toName="context"/>
+</View>"""
+    }
+
+    response = client.post('/predict', data=json.dumps(request), content_type='application/json')
+    assert response.status_code == 200
```
Comment (niklub) [inline label_studio_ml/examples/watsonx_llm/test_api.py:77]: how does it work? does it use live request to ibm?

**Review by niklub — CHANGES_REQUESTED** [review]

File: label_studio_ml/examples/watsonx_llm/README.md (hunk 1 above)
Comment (micaelakaplan) [reply label_studio_ml/examples/watsonx_llm/README.md:3]: Added below!

File: label_studio_ml/examples/watsonx_llm/data_transfer_app.py (hunk 5 above)
Comment (micaelakaplan) [reply label_studio_ml/examples/watsonx_llm/data_transfer_app.py:39]: Can't set defaults, as these values are unique to each user, but set to throw an exception if they are not provided.

File: label_studio_ml/examples/watsonx_llm/model.py (hunk 11 above)
Comment (micaelakaplan) [reply label_studio_ml/examples/watsonx_llm/model.py:62]: tableing for a separate PR

File: label_studio_ml/examples/watsonx_llm/model.py (hunk 10 above)
Comment (micaelakaplan) [reply label_studio_ml/examples/watsonx_llm/model.py]: resolved to not assume reinitialization

File: label_studio_ml/examples/watsonx_llm/test_api.py (hunk 14 above)
Comment (micaelakaplan) [reply label_studio_ml/examples/watsonx_llm/test_api.py:77]: Yes -- need to update this, but will save for a future PR

**Review by niklub — APPROVED** [review]
LGTM, feel free to merge after fixing the test
